lib/pure/os.nim

```diff
@@ -1647,27 +1647,94 @@ proc setFilePermissions*(filename: string, permissions: set[FilePermission]) {.
       var res2 = setFileAttributesA(filename, res)
     if res2 == - 1'i32: raiseOSError(osLastError(), $(filename, permissions))
 
-proc copyFile*(source, dest: string) {.rtl, extern: "nos$1",
-  tags: [ReadIOEffect, WriteIOEffect], noWeirdTarget.} =
+proc createSymlink*(src, dest: string) {.tags: [ReadDirEffect, WriteIOEffect],
+  gcsafe, noWeirdTarget.} =
+  ## Create a symbolic link at `dest` which points to the item specified
+  ## by `src`. On most operating systems, will fail if a link already exists.
+  ##
+  ## **Warning**:
+  ## Some OS's (such as Microsoft Windows) restrict the creation
+  ## of symlinks to root users (administrators).
+  ##
+  ## See also:
+  ## * `createHardlink proc <#createHardlink,string,string>`_
+  ## * `expandSymlink proc <#expandSymlink,string>`_
+
+  when defined(Windows):
+    # 2 is the SYMBOLIC_LINK_FLAG_ALLOW_UNPRIVILEGED_CREATE. This allows
+    # anyone with developer mode on to create a link
+    let flag = dirExists(src).int32 or 2
+    when useWinUnicode:
+      var wSrc = newWideCString(src)
+      var wDst = newWideCString(dest)
+      if createSymbolicLinkW(wDst, wSrc, flag) == 0 or getLastError() != 0:
+        raiseOSError(osLastError(), $(src, dest))
+    else:
+      if createSymbolicLinkA(dest, src, flag) == 0 or getLastError() != 0:
+        raiseOSError(osLastError(), $(src, dest))
+  else:
+    if symlink(src, dest) != 0:
+      raiseOSError(osLastError(), $(src, dest))
+
+proc expandSymlink*(symlinkPath: string): string {.tags: [ReadIOEffect],
+  gcsafe, noWeirdTarget.} =
+  ## Returns a string representing the path to which the symbolic link points.
+  ##
+  ## On Windows this is a noop, ``symlinkPath`` is simply returned.
+  ##
+  ## See also:
+  ## * `createSymlink proc <#createSymlink,string,string>`_
+  when defined(windows):
+    result = symlinkPath
+  else:
+    result = newString(maxSymlinkLen)
+    var len = readlink(symlinkPath, result, maxSymlinkLen)
+    if len < 0:
+      raiseOSError(osLastError(), symlinkPath)
+    if len > maxSymlinkLen:
+      result = newString(len+1)
+      len = readlink(symlinkPath, result, len)
+    setLen(result, len)
+
+type
+  ## Enumeration specifying an action to perform on symlink while copying.
+  SymlinkAction* = enum
+    saFollow,             ## Copy the files symlinks point to
+    saCopy,               ## Copy symlinks as symlinks
+    saSkip                ## Ignore symlinks
+
+  ## Options controlling the behavior of copying files.
+  CopyOptions* = object
+    symlinkAction*: SymlinkAction  ## Action to perform on symlink
+
+const copyFileCopyOptions* = CopyOptions(symlinkAction: saFollow)
+
+proc copyFile*(source, dest: string, options = copyFileCopyOptions) {.rtl,
```

Comment:
how about inline the default options: more readable, less bloat
```nim
proc copyFile*(source, dest: string, options = CopyOptions(symlinkAction: saFollow)) 
```

Reply:
I've wanted to avoid extra object creation. Or is it something that Nim compiler optimizes for me?

Reply:
Anyway, I do agree that it's much more readable to inline them, so I've updated my commit.

Reply:
> I've wanted to avoid extra object creation. Or is it something that Nim compiler optimizes for me?

@rominf no extra object creation cost, but see https://github.com/nim-lang/Nim/issues/16738